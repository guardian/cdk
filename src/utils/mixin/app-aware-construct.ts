import { Construct } from "constructs";
import { AppIdentity } from "../../constructs/core";
import type { AnyConstructor } from "./types";

export function GuAppAwareConstruct<TBase extends AnyConstructor>(BaseClass: TBase) {
  class Mixin extends BaseClass {
    /**
     * The ID of the construct with the App suffix.
     * This should be used in place of `id` when trying to reference the construct.
     */
    readonly idWithApp: string;

    // eslint-disable-next-line custom-rules/valid-constructors, @typescript-eslint/no-explicit-any -- mixin
    protected constructor(...args: any[]) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- mixin
      const [scope, id, props, ...rest] = args;

      if (!AppIdentity.isAppIdentity(props)) {
        throw new Error("Cannot use the GuAppAwareConstruct mixin without an AppIdentity");
      }

      const app: string = props.app;
      const idWithApp = AppIdentity.suffixText({ app }, id as string);

      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- mixin
      const newArgs = [scope, idWithApp, props, ...rest];
      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument -- mixin
      super(...newArgs);

      this.idWithApp = idWithApp;

      /*
        Add the `App` tag to the construct.
        Although not every resource can be tagged, it's still safe to make the call.
        If AWS support tags on a new resource one day, our test suite will fail and we can celebrate!
        See https://docs.aws.amazon.com/ARG/latest/userguide/supported-resources.html
         */
      if (Construct.isConstruct(this)) {
        AppIdentity.taggedConstruct({ app }, this);
      }
    }
  }
  return Mixin;
}
