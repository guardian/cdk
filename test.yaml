Parameters:
  Stage:
    Type: String
    Default: CODE
    AllowedValues:
      - CODE
      - PROD
    Description: Stage name
Resources:
  accountservicesanghammaradtopicarnEBB061EC:
    Type: AWS::SSM::Parameter
    Properties:
      Type: String
      Value: cdk-update
      Description: NIC Test
      Name: /NIC/test
      Tags:
        gu:cdk:version: 35.0.1
        gu:repo: guardian/cdk
        Stack: foo
        Stage:
          Ref: Stage
