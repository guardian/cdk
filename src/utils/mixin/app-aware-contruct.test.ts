import { Template } from "aws-cdk-lib/assertions";
import type { BucketProps } from "aws-cdk-lib/aws-s3";
import { Bucket } from "aws-cdk-lib/aws-s3";
import type { AppIdentity, GuStack } from "../../constructs/core";
import { simpleGuStackForTesting } from "../test";
import { GuAppAwareConstruct } from "./app-aware-construct";

// `GuAppAwareConstruct` should only operate if `props` has an `app` property,
// so usage here should be a no-op.
class TestConstruct extends GuAppAwareConstruct(Bucket) {
  constructor(scope: GuStack, id: string, props: BucketProps) {
    super(scope, id, props);
  }
}

interface TestAppAwareConstructProps extends BucketProps, AppIdentity {}

class TestAppAwareConstruct extends GuAppAwareConstruct(Bucket) {
  constructor(scope: GuStack, id: string, props: TestAppAwareConstructProps) {
    super(scope, id, props);
  }
}

describe("The GuAppAwareConstruct mixin", () => {
  // demonstrates usage of `GuAppAwareConstruct`
  it("should throw if no app identifier is provided", () => {
    const stack = simpleGuStackForTesting();

    expect(() => {
      new TestConstruct(stack, "MyBucket", {});
    }).toThrowError("Cannot use the GuAppAwareConstruct mixin without an AppIdentity");
  });

  it("should suffix the id with the app identifier", () => {
    const stack = simpleGuStackForTesting();
    const bucket = new TestAppAwareConstruct(stack, "MyBucket", { app: "Test" });

    expect(bucket.idWithApp).toBe("MyBucketTest");
  });

  it("should add the app tag", () => {
    const stack = simpleGuStackForTesting();
    new TestAppAwareConstruct(stack, "MyBucket", { app: "Test" });
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });
});
