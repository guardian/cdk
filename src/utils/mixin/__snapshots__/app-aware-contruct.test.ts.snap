// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuAppAwareConstruct mixin should add the app tag 1`] = `
Object {
  "Resources": Object {
    "MyBucketTest262E966E": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "Test",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
